server/controllers/manage/premises/placements/changesController.ts

```diff
@@ -0,0 +1,261 @@
+        backlink: adminPaths.admin.placementRequests.show({ id: placement.requestForPlacementId }),
+        pageHeading: 'Change placement',
+        placement,
+        startDate,
+        ...DateFormats.isoDateToDateInputs(startDate, 'startDate'),
+        durationDays,
+        criteria,
+        placementSummary: placementOverviewSummary(placement),
+        placementDatesSummary: placementDatesSummary(placement),
+        durationOptions: durationSelectOptions(durationDays),
+        criteriaOptions: convertKeyValuePairToCheckBoxItems(occupancyCriteriaMap, criteria),
+        summary,
+        calendar,
+        errors,
+        errorSummary,
+        ...startDateParts,
+        ...userInput,
+      })
+    }
+  }
+
+  saveNew(): RequestHandler {
+    return async (req: ViewRequest, res: Response) => {
+      const { body, params } = req
+
+      try {
+        const errors = validateSpaceBooking(body)
+
+        if (Object.keys(errors).length) {
+          throw new ValidationError(errors)
+        }
+
+        const { arrivalDate } = DateFormats.dateAndTimeInputsToIsoString(req.body, 'arrivalDate')
+        const { departureDate } = DateFormats.dateAndTimeInputsToIsoString(req.body, 'departureDate')
+        const { criteria } = req.body
+
+        const redirectUrl = `${managePaths.premises.placements.changes.confirm(params)}?${createQueryString(
+          {
+            arrivalDate,
+            departureDate,
+            criteria: criteria.split(','),
+          },
+          { arrayFormat: 'repeat' },
+        )}`
+
+        return res.redirect(redirectUrl)
+      } catch (error) {
+        const viewUrl = `${managePaths.premises.placements.changes.new(params)}${createQueryString(req.query, {
+          arrayFormat: 'repeat',
+          addQueryPrefix: true,
+        })}`
+        return catchValidationErrorOrPropogate(req, res, error, viewUrl)
+      }
+    }
+  }
+
+  confirm(): RequestHandler {
+    return async (req: ConfirmRequest, res: Response) => {
+      const { token } = req.user
+      const { premisesId, placementId } = req.params
+      const { arrivalDate, departureDate, criteria } = req.query
+
+      const [premises, placement] = await Promise.all([
+        this.premisesService.find(token, premisesId),
+        this.placementService.getPlacement(token, placementId),
+      ])
+
+      const backlink = `${managePaths.premises.placements.changes.new(req.params)}${createQueryString(
+        {
+          criteria,
+          arrivalDate,
+          departureDate,
+        },
+        { arrayFormat: 'repeat', addQueryPrefix: true },
+      )}`
+
+      return res.render('manage/premises/placements/changes/confirm', {
+        pageHeading: 'Confirm booking changes',
+        backlink,
+        placement,
+        summaryListRows: spaceBookingConfirmationSummaryListRows(
+          premises,
+          arrivalDate,
+          departureDate,
+          makeArrayOfType<Cas1SpaceBookingCharacteristic>(criteria),
```

Comment:
This needs a || [] as makeArrayOfType returns undefined if criteria is falsey and requirementsHtmlString then blows up

Reply:
Fixed!